src/hotspot/share/c1/c1_GraphBuilder.cpp

```diff
@@ -3452,6 +3463,17 @@ bool GraphBuilder::try_inline(ciMethod* callee, bool holder_known, bool ignore_r
   return false;
 }
 
+bool GraphBuilder::try_blackhole(ciMethod* callee) {
+  Values* args = state()->pop_arguments(callee->arg_size());
+
+  // Blackhole everything except the receiver itself
```

Comment:
Does it make sense to limit blackhole candidates to static methods only? 
Otherwise, you end up with a null check in the generated code.

Reply:
Unfortunately, not. In JMH, `Blackhole` methods are instance methods (they have state), and that plays nicely with compatibility: when JIT fails to provide blackholing, it would naturally fall back to old behavior. I think null check for receiver is the fair behavior here, as subsequent optimization may expect receiver null-checked after the call. I have not seen null-checks in my `perfasm` runs with C1, though.